### PR TITLE
clearing old held outputs

### LINKED ISSUE #295: Memory leak in "held" output implementation of ensemble model
--------
Issue Description
--------
##### What did you expect to see happen?
Flat memory growth after initialization

##### What did you see instead?
Significant mostly-linear memory usage growth

##### Do you have a suggested fix for the development team?
Not enough "old" held outputs are being cleared, causing wasteful storage

----------------
For Change Control Board: Issue Review
----------------
This review should occur before any development is performed as a response to this issue.
- [x] 1. Is it tagged with a type: defect or improvement?
- [x] 2. Is it tagged with a priority: critical, normal or minor?
- [x] 3. If it will impact requirements or requirements tests, is it tagged with requirements?
- [x] 4. If it is a defect, can it cause wrong results for users? If so an email needs to be sent to the users.
- [x] 5. Is a rationale provided? (Such as explaining why the improvement is needed or why current code is wrong.)

-------
For Change Control Board: Issue Closure
-------
This review should occur when the issue is imminently going to be closed.
- [ ] 1. If the issue is a defect, is the defect fixed?
- [ ] 2. If the issue is a defect, is the defect tested for in the regression test system?  (If not explain why not.)
- [ ] 3. If the issue can impact users, has an email to the users group been written (the email should specify if the defect impacts stable or master)?
- [ ] 4. If the issue is a defect, does it impact the latest stable branch? If yes, is there any issue tagged with stable (create if needed)?
- [ ] 5. If the issue is being closed without a merge request, has an explanation of why it is being closed been provided?

### PR DESCRIPTION
**Pull request by PaulTalbot-INL**
--------
Pull Request Description
--------
##### What issue does this change request address? (Use "#" before the issue to link it, i.e., #42.)
Closes #295.

##### What are the significant changes in functionality due to this change request?
Patches (not truly fixes) the storage problem mentioned in issue #295.

I discovered that when a potential new optimal point is rejected, it somehow fails to trigger old output removal until a new optimal point is accepted.  The patch is to clear all old outputs that are held once a new point is accepted.

----------------
For Change Control Board: Change Request Review
----------------
The following review must be completed by an authorized member of the Change Control Board.
- [x] 1. Review all computer code.
- [x] 2. If any changes occur to the input syntax, there must be an accompanying change to the user manual and xsd schema. If the input syntax change deprecates existing input files, a conversion script needs to be added (see Conversion Scripts).
- [x] 3. Make sure the Python code and commenting standards are respected (camelBack, etc.) - See on the [wiki](https://github.com/idaholab/raven/wiki/RAVEN-Code-Standards#python) for details.
- [x] 4. Automated Tests should pass, including run_tests, pylint, manual building and xsd tests. If there are changes to Simulation.py or JobHandler.py the qsub tests must pass.
- [x] 5. If significant functionality is added, there must  be tests added to check this. Tests should cover all possible options.  Multiple short tests are preferred over one large test. If new development on the internal JobHandler parallel system is performed, a cluster test must be added setting, in <RunInfo> XML block, the node ```<internalParallel>``` to True.
- [x] 6. If the change modifies or adds a requirement or a requirement based test case, the Change Control Board's Chair or designee also needs to approve the change.  The requirements and the requirements test shall be in sync.
- [x] 7. The merge request must reference an issue.  If the issue is closed, the issue close checklist shall be done.
- [x] 8. If an analytic test is changed/added is the the analytic documentation updated/added?

### CODE REVIEW

File: framework/Models/EnsembleModel.py [hunk 1]
```diff
@@ -467,12 +467,20 @@ def collectOutput(self,finishedJob,output):
...
-        keys = self.tempOutputs['forHold'].keys()
-        for key in keys:
-          if key.startswith(exportDict['metadata']['holdOutputErase'][0]):
-            self.tempOutputs['forHold'].pop(key)
+    # first clear old outputs
+    # TODO FIXME this is a flawed implementation, since it requires that the "holdOutputErase" is of
+    # a very specific form that works with the the current SPSA optimizer.  Since we have no other optimizer right now,
+    # the problem is only extensibility, not the actual implementation.
+    if exportDict['metadata'].get('holdOutputErase',None) is not None:
+      #if exportDict['metadata']['holdOutputErase'] is not None:
```
Comment (AaronEpiney) [inline framework/Models/EnsembleModel.py]: Is code commenting OK? Does this have to be removed?

File: framework/Models/EnsembleModel.py [hunk 2]
```diff
@@ -467,12 +467,20 @@ def collectOutput(self,finishedJob,output):
...
+      keys = self.tempOutputs['forHold'].keys()
+      toErase = exportDict['metadata']['holdOutputErase'][0].split('_')[:2]
+      for key in keys:
+        traj,itr,pert = key.split('_')
+        if traj == toErase[0] and itr <= toErase[1]:
+          self.tempOutputs['forHold'].pop(key)
```
Comment (AaronEpiney) [inline framework/Models/EnsembleModel.py]: Is popping elements the RAVEN standard? Not del?

File: framework/Models/EnsembleModel.py (hunk 1 above)
Comment (PaulTalbot-INL) [inline framework/Models/EnsembleModel.py]: I'll remove it.

File: framework/Models/EnsembleModel.py (hunk 2 above)
Comment (PaulTalbot-INL) [inline framework/Models/EnsembleModel.py]: I like "del" better, I'll make that change.

Comment (PaulTalbot-INL) [inline framework/Models/EnsembleModel.py]: For the record, I ran a script to test the timing of these two operations:
```python
from timeit import timeit as t

def const():
  return {'b':{'c':6}}

def runDel():
  a = const()
  del a['b']['c']

def runPop():
  a = const()
  a['b'].pop('c')

d = t(runDel,number=1000000)
p = t(runPop,number=1000000)

print 'Seconds for del:',d
print 'Seconds for pop:',p
```
The results on my macbook was:
```
Seconds for del: 0.488843917847
Seconds for pop: 0.568461179733
```

**Pull request merged**